### PR TITLE
Fix returning the proper last inserted id in MSSQL

### PR DESCRIPTION
**Pull request by sarven**
|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | yes
| Fixed issues | #3516 

#### Summary
I discovered a problem when the table in MSSQL has a trigger with insert statements. Id returned by method lastInsertId is invalid. It isn't id from the table in where we were inserting data, but from the table that trigger was inserting data.

The problem concerns sqlsrv and pdo_sqlsrv extensions. The fix for sqlsrv was quite simple, I only added a few lines of code. 

But the fix for pdo_sqlsrv required to rebuild the implementation of Connection and Statement. I made it similar to the sqlsrv implementation.

### CODE REVIEW

File: tests/Doctrine/Tests/DBAL/Functional/Driver/PDOSqlsrv/DriverTest.php [hunk 1]
```diff
@@ -64,11 +63,4 @@ public function testConnectionOptions() : void
 
         self::assertSame('APP_NAME', $result);
     }
-
-    public function testDriverOptions() : void
```
Comment (SenseException) [inline tests/Doctrine/Tests/DBAL/Functional/Driver/PDOSqlsrv/DriverTest.php]: Is this test breaking with your changes?

Comment (sarven) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/PDOSqlsrv/DriverTest.php]: Yes, because previously Connection extended PDOConnection, now it only implements Connection interface, where is no method like getAttribute.

Comment (SenseException) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/PDOSqlsrv/DriverTest.php]: master-branch is currently for 2.10 while develop is for 3.0. It looks like the changes are more fitting for the next major release.

Comment (sarven) [reply tests/Doctrine/Tests/DBAL/Functional/Driver/PDOSqlsrv/DriverTest.php]: So should I make a branch based on develop then apply these changes and create a new pull request or just override this?

**Comment by morozov** [comment]
@sarven could you rebase this please?

**Comment by sarven** [comment]
@morozov Done.

File: .appveyor.yml [hunk 2]
```diff
@@ -68,7 +68,7 @@ install:
           Add-Content php.ini "`n extension=php_curl.dll"
           Add-Content php.ini "`n curl.cainfo=C:\tools\cacert\bundle.pem"
 
-          $DLLVersion = "5.5.0preview"
+          $DLLVersion = "5.6.1"
```
Comment (morozov) [inline .appveyor.yml]: Please revert unless this is required. If possible, we should roll back to using the latest stable version as it was prior to #3484 but that would be a different PR.

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php [hunk 3]
```diff
@@ -1,37 +0,0 @@
...
-namespace Doctrine\DBAL\Driver\PDOSqlsrv;
-
-use Doctrine\DBAL\Driver\PDOStatement;
-use Doctrine\DBAL\ParameterType;
-use PDO;
-
-/**
- * PDO SQL Server Statement
- */
-class Statement extends PDOStatement
```
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php]: Why does this need to be renamed?

**Review by morozov — COMMENTED** [review]
@sarven thank you. Could you please describe the problem in the pull request description?

File: tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ */
+class DBAL3516Test extends DbalFunctionalTestCase
+{
+    protected function setUp() : void
+    {
+        parent::setUp();
+
+        $platform = $this->connection->getDatabasePlatform()->getName();
+
+        if ($platform !== 'mssql') {
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php]: Since this is not the only driver/platform which implements `lastInsertId()`, the test should be implemented for all supported platforms and skipped if there's a known issue with a specific platform.

File: tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+declare(strict_types=1);
+
+namespace Doctrine\Tests\DBAL\Functional\Ticket;
+
+use Doctrine\Tests\DbalFunctionalTestCase;
+
+/**
+ * @group DBAL-3516
+ */
+class DBAL3516Test extends DbalFunctionalTestCase
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php]: Please move it under `Doctrine\Tests\DBAL\Functional\ConnectionTest`.

File: tests/Doctrine/Tests/DBAL/Functional/StatementTest.php [hunk 6]
```diff
@@ -373,7 +374,9 @@ public function testExecWithRedundantParameters() : void
      */
     public function testFetchColumnNonExistingIndex(int $index) : void
     {
-        if ($this->connection->getWrappedConnection() instanceof PDOConnection) {
+        $wrappedConnection = $this->connection->getWrappedConnection();
+
+        if ($wrappedConnection instanceof PDOConnection || $wrappedConnection instanceof PDOSqlsrvConnection) {
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/StatementTest.php]: Why isn't `PDOSqlsrvConnection` sub-type of `PDOConnection`?

File: .appveyor.yml (hunk 2 above)
Comment (sarven) [reply .appveyor.yml]: @morozov I reverted, but probably some tests won't pass.

File: tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php (hunk 4 above)
Comment (sarven) [reply tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php]: I thought about make this tests for all platforms, but there is a problem with create trigger for all platforms. So I think that creating it only for MSSQL is ok, because this problem probably doesn't concern others platforms.

File: tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php (hunk 5 above)
Comment (sarven) [reply tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php]: Ok

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php (hunk 3 above)
Comment (sarven) [reply lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php]: I just wanted to name it similarly to SQLSrvStatement. It is not subtype of PDOStatement, so I wanted to mark it clearly.

File: tests/Doctrine/Tests/DBAL/Functional/StatementTest.php (hunk 6 above)
Comment (sarven) [reply tests/Doctrine/Tests/DBAL/Functional/StatementTest.php]: I did some changes and now Connection and Statement in PDOSqlsrv are subtypes of specific doctrine PDO classes. Previously I wanted it to be similarly to the Sqlsrv implementation, but it is also possible to just extends PDO classes and do some overrides. Now it seems to be simpler.

File: .appveyor.yml (hunk 2 above)
Comment (sarven) [reply .appveyor.yml]: @morozv Now in appveyor is the error:
`
PHP Fatal error:  Meta data vector out of sync in
`
Should I create a new pull request with a change of version?

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php [hunk 7]
```diff
@@ -14,19 +17,31 @@
  */
 class Connection extends PDOConnection
 {
+    /** @var LastInsertId|null */
+    protected $lastInsertId;
+
+    /**
+     * Append to any INSERT query to retrieve the last insert id.
+     */
+    protected const LAST_INSERT_ID_SQL = ';SELECT SCOPE_IDENTITY() AS LastInsertId;';
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php]: `private` unless needed elsewhere

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php [hunk 8]
```diff
@@ -14,19 +17,31 @@
  */
 class Connection extends PDOConnection
 {
+    /** @var LastInsertId|null */
+    protected $lastInsertId;
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php]: `private` unless needed elsewhere

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Driver\PDOSqlsrv;
+
+/**
+ * Last Id Data Container.
+ */
+class LastInsertId
+{
+    /** @var string */
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php]: According to (missing) constructor semantics, this is `string|null`

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    /**
+     * @param string $id
+     */
+    public function setId($id)
+    {
+        $this->id = $id;
+    }
+
+    /**
+     * @return string
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php]: According to (missing) constructor semantics, this is `string|null`

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ */
+class LastInsertId
+{
+    /** @var string */
+    private $id;
+
+    /**
+     * @param string $id
+     */
+    public function setId($id)
```
Comment (Ocramius) [inline lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php]: Type declarations over docblocks: this is `string` and `: void`. Same for the rest of the patch overall.

**Review by Ocramius — CHANGES_REQUESTED** [review]
I'd need @deeky666 to check this patch, since he initially authored https://github.com/doctrine/dbal/pull/2648 (which was closed due to its high complexity/risk for little gain).

I'd say that also here the same applies: if a prepared statement is causing side-effects through stored procedures and triggers, I'd rather call the schema "foobar'd" and "needs to be fixed there" rather than trying to fix issues on this end.

**Comment by warslett** [comment]
hey @sarven @Ocramius any update on this PR? looks like the branches might need clearing up (presuming this PR doesn't include 141 commits).
